pkg/ingester/user_state.go

```diff
@@ -353,3 +319,65 @@ outer:
 	}
 	return nil
 }
+
+const numMetricCounterShards = 128
+
+type metricCounterShard struct {
+	mtx sync.Mutex
+	m   map[string]int
+}
+
+type metricCounter struct {
+	limiter *Limiter
+	shards  []metricCounterShard
+}
+
+func newMetricCounter(limiter *Limiter) *metricCounter {
+	shards := make([]metricCounterShard, 0, numMetricCounterShards)
+	for i := 0; i < numMetricCounterShards; i++ {
+		shards = append(shards, metricCounterShard{
+			m: map[string]int{},
+		})
+	}
+	return &metricCounter{
+		limiter: limiter,
+		shards:  shards,
+	}
+}
+
+func (m *metricCounter) removeMetricName(metricName string) {
+	shard := m.getShard(metricName)
+	shard.mtx.Lock()
+	defer shard.mtx.Unlock()
+
+	shard.m[metricName]--
+	if shard.m[metricName] == 0 {
+		delete(shard.m, metricName)
+	}
+}
+
+func (m *metricCounter) getShard(metricName string) *metricCounterShard {
+	shard := &m.shards[util.HashFP(model.Fingerprint(fnv1a.HashString64(metricName)))%numMetricCounterShards]
+	return shard
+}
+
+func (m *metricCounter) canAddSeriesFor(userID, metric string) error {
+	shard := m.getShard(metric)
+	shard.mtx.Lock()
+	defer shard.mtx.Unlock()
+
+	err := m.limiter.AssertMaxSeriesPerMetric(userID, shard.m[metric])
+	if err != nil {
+		return err
+	}
+
+	shard.m[metric]++
```

Comment:
Why do we increment it here, if then we also call `seriesAddedFor()`? This means we're double tracking it. Don't we have any test covering the in-memory series limit? If not, we should add it.

Reply:
This actually should not be here, a copy paste error